contrib/shibuqam/examples/shibuqamoauth.nit

```diff
@@ -0,0 +1,157 @@
+# * `avatar`: the gravatar based on `email`
+# * `state`: the same state to check that the response is not forged.
+#
+# ## Errors
+#
+# If the request is badly formed or if the `redir` is not allowed, then there is no redirection and a text error message is send to the user.
+#
+# If there is a problem during the authentication, then there is a redirection but the GET fields are only `state` and `error`, where `error` is an error message.
+#
+# # To configure and execute
+#
+# ## Run the server
+#
+#    shibuqamoauth authorized_list [host] [port]
+#
+# `authorized_list` is a textfile that registers the list of accepted `redir`.
+#
+# Example:
+#
+# ~~~raw
+# https://example.com/foo
+# https://other.example.com/
+# ~~~
+#
+# To be authorized, a `redir` must have one of the line of `authorized_list` as a prefix.
+# For instance, the previous `authorized_list` accepts `https://example.com/foo` and `https://example.com/foo/bar`
+# but not `https://example.com/bar` nor `https://sub.example.com/foo`.
+#
+#
+# ## Install as a server
+#
+# A correct `shibuqam` reverse proxy must be configured (see `shibuqam` for details).
+# A full scenario is the following:
+#
+# * User: the human user and its browser that use the client website.
+# * Client: the web site that need authentication. It owns `https://client.example.com`
+# * Proxy: the public reverse proxy that does genuine Shibboleth authentication. It owns `https://server.example.com`
+# * Service: the private shibuqamoauth service. It owns `https://shibuqamoauth.example.com/`
+#
+# 1. User wants to login on client.
+# 2. Client chooses a random secret (state).
+#    Redirects the user to proxy as `https://server.example.com/login?redir=https://client.example.com/callback&state=secret`
+# 3. Proxy does the Shibboleth authentication.
+#    Enhances the request header.
+#    Forwards the request to service as `https://shibuqamoauth.example.com/login?redir=https://client.example.com/callback&state=secret`
+# 4. Service checks that `redir` is authorized.
+#    Digs in the enhanced request header to build a callback.
+#    Redirect the user to client as `https://client.example.com/callback?id=jdoe&name=John+Doe&email=foo&avatar=bar&state=secret`
+# 5. Client checks that the `state` in the callback matches the one in the initial redirection.
+#    Trusts the user information from the provided GET fields.
+module shibuqamoauth
+
+import popcorn
+import shibuqam
+
+class AuthHandler
+	super Handler
+
+	# List of prefixes of authorized redirections.
+	var authorized: Array[String]
+
+	redef fun get(http_request, response)
+	do
+		var redir = http_request.get_args.get_or_null("redir")
+		if redir == null then
+			response.send("No redir.", 400)
+			return
+		end
+		redir = redir.from_percent_encoding
+		var good = false
+		for r in authorized do
+			if redir.has_prefix(r) then
+				good = true
+				break
+			end
+		end
+		if not good then
+			response.send("Site not authorized.", 403)
+			return
+		end
```

Comment:
Especially for security-related checks, it is a good idea to split a big routine into small functions so it is easier to unit-test and review each chunk individually then conclude that the whole routine is right. For example, here, lines 104–114 should be extracted into an auxiliary function.

Please divide, so we can conquer. :wink: